File: flutter/packages/workout_log_core/lib/src/cycles/cycle_templates.dart
import '../models/cycle.dart';
import 'cycle_day.dart';

/// The block skeleton a new workout starts from, chosen by its number.
///
/// Lifted from the shape of the hybrid-8 cycle rather than invented: a `1` day
/// is a conditioning day (short warm-up, explosive lift, metcon) and a `2` day
/// is a heavy day (longer warm-up, hyperextension, a main lift, accessories).
/// Every field stays editable afterwards.
abstract final class CycleTemplates {
  static List<BlockTemplate> blocksFor(DayKind kind) => switch (kind) {
        DayKind.conditioning => [
            BlockTemplate(type: 'cardio', role: 'warmup', machine: '', durationMin: 10),
            BlockTemplate(type: 'strength', role: 'explosive'),
            BlockTemplate(type: 'metcon', role: 'metcon'),
            BlockTemplate(type: 'strength', role: 'accessory', setsReps: [6, 6, 6, 6]),
            BlockTemplate(type: 'strength', role: 'grip', setsReps: [8, 8, 8, 8]),
            BlockTemplate(type: 'cooldown', role: 'cooldown'),
          ],
        DayKind.heavy => [
            BlockTemplate(type: 'cardio', role: 'warmup', machine: '', durationMin: 15),
            BlockTemplate(
              type: 'strength',
              role: 'warmup',
              exercise: 'гіперекстензія',
              setsReps: [12, 12],
            ),
            BlockTemplate(type: 'strength', role: 'main'),
            BlockTemplate(type: 'strength', role: 'accessory', setsReps: [6, 6, 6, 6, 6]),
            BlockTemplate(type: 'strength', role: 'accessory', setsReps: [6, 6, 6, 6, 6]),
            BlockTemplate(type: 'strength', role: 'grip', setsReps: [8, 8, 8, 8]),
            BlockTemplate(type: 'cooldown', role: 'cooldown'),
          ],
      };

  /// A fresh workout for `day`, pre-filled and slotted after `previous`.
  static CycleSession session(CycleDay day, {int? week, String? weekday}) =>
      CycleSession(
        cycleDay: day.code,
        week: week,
        weekday: weekday,
        type: day.kind == DayKind.conditioning ? 'metcon' : 'heavy',
        blocks: CycleTemplates.blocksFor(day.kind),
      );

  /// The code a new workout should get: one past the highest that follows the
  /// convention, or `A1` when the cycle is empty or entirely free-form.
  static CycleDay nextDay(Iterable<String> existingCodes) {
    final parsed = existingCodes
        .map(CycleDay.tryParse)
        .whereType<CycleDay>()
        .toList()
      ..sort();
    return parsed.isEmpty ? const CycleDay('A', DayKind.conditioning) : parsed.last.next;
  }
}

File: crates/workout-log-core/src/cycles/cycle_templates.rs
use super::cycle_day::{CycleDay, DayKind};
use crate::models::cycle::{BlockTemplate, CycleSession};

fn block(block_type: &str, role: &str) -> BlockTemplate {
    BlockTemplate {
        block_type: block_type.into(),
        role: Some(role.into()),
        machine: None,
        duration_min: None,
        exercise: None,
        sets_reps: None,
        notes: None,
        format: None,
        scheme: None,
        exercises: None,
        extras: Default::default(),
    }
}

fn cardio(role: &str, minutes: f64) -> BlockTemplate {
    BlockTemplate {
        machine: Some(String::new()),
        duration_min: Some(minutes),
        ..block("cardio", role)
    }
}

fn strength(role: &str, exercise: Option<&str>, sets_reps: Option<Vec<i64>>) -> BlockTemplate {
    BlockTemplate {
        exercise: exercise.map(str::to_string),
        sets_reps,
        ..block("strength", role)
    }
}

pub fn blocks_for(kind: DayKind) -> Vec<BlockTemplate> {
    match kind {
        DayKind::Conditioning => vec![
            cardio("warmup", 10.0),
            strength("explosive", None, None),
            block("metcon", "metcon"),
            strength("accessory", None, Some(vec![6, 6, 6, 6])),
            strength("grip", None, Some(vec![8, 8, 8, 8])),
            block("cooldown", "cooldown"),
        ],
        DayKind::Heavy => vec![
            cardio("warmup", 15.0),
            strength("warmup", Some("гіперекстензія"), Some(vec![12, 12])),
            strength("main", None, None),
            strength("accessory", None, Some(vec![6, 6, 6, 6, 6])),
            strength("accessory", None, Some(vec![6, 6, 6, 6, 6])),
            strength("grip", None, Some(vec![8, 8, 8, 8])),
            block("cooldown", "cooldown"),
        ],
    }
}

pub fn session(day: CycleDay, week: Option<i64>, weekday: Option<String>) -> CycleSession {
    CycleSession {
        cycle_day: day.code(),
        week,
        weekday,
        session_type: Some(
            match day.kind {
                DayKind::Conditioning => "metcon",
                DayKind::Heavy => "heavy",
            }
            .into(),
        ),
        title: None,
        session_notes: None,
        blocks: blocks_for(day.kind),
        extras: Default::default(),
    }
}

/// The next code in the A1/A2 progression after everything already planned.
pub fn next_day(existing_codes: &[String]) -> CycleDay {
    let mut days: Vec<CycleDay> = existing_codes
        .iter()
        .filter_map(|c| CycleDay::try_parse(c))
        .collect();
    days.sort_by_key(|d| d.code());
    days.last().map_or(
        CycleDay {
            letter: 'A',
            kind: DayKind::Conditioning,
        },
        |last| last.next(),
    )
}
